CUDADataFormats/Track/interface/PixelTrackUtilities.h

```diff
@@ -0,0 +1,241 @@
+#ifndef CUDADataFormats_Track_PixelTrackUtilities_h
+#define CUDADataFormats_Track_PixelTrackUtilities_h
+
+#include <Eigen/Dense>
+#include <Eigen/Core>
+#include "Geometry/CommonTopologies/interface/SimplePixelTopology.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/HistoContainer.h"
+#include "DataFormats/SoATemplate/interface/SoALayout.h"
+
+namespace pixelTrackSoA {
+
+  enum class Quality : uint8_t { bad = 0, edup, dup, loose, strict, tight, highPurity, notQuality };
+  constexpr uint32_t qualitySize{uint8_t(Quality::notQuality)};
+  const std::string qualityName[qualitySize]{"bad", "edup", "dup", "loose", "strict", "tight", "highPurity"};
+  inline Quality qualityByName(std::string const &name) {
+    auto qp = std::find(qualityName, qualityName + qualitySize, name) - qualityName;
+    return static_cast<Quality>(qp);
+  }
+
+}  // namespace pixelTrackSoA
+
+template <typename TrackerTraits>
+struct TrackSoA {
+  static constexpr int32_t S = TrackerTraits::maxNumberOfTuples;
+  static constexpr int32_t H = TrackerTraits::avgHitsPerTrack;
+  // Aliases in order to not confuse the GENERATE_SOA_LAYOUT
+  // macro with weird colons and angled brackets.
+  using Vector5f = Eigen::Matrix<float, 5, 1>;
+  using Vector15f = Eigen::Matrix<float, 15, 1>;
+  using Quality = pixelTrackSoA::Quality;
+
+  using hindex_type = uint32_t;
+
+  using HitContainer = cms::cuda::OneToManyAssoc<hindex_type, S + 1, H * S>;
+
+  GENERATE_SOA_LAYOUT(TrackSoALayout,
+                      SOA_COLUMN(Quality, quality),
+                      SOA_COLUMN(float, chi2),
+                      SOA_COLUMN(int8_t, nLayers),
+                      SOA_COLUMN(float, eta),
+                      SOA_COLUMN(float, pt),
+                      SOA_EIGEN_COLUMN(Vector5f, state),
+                      SOA_EIGEN_COLUMN(Vector15f, covariance),
+                      SOA_SCALAR(int, nTracks),
+                      SOA_SCALAR(HitContainer, hitIndices),
+                      SOA_SCALAR(HitContainer, detIndices))
+};
+
+// Methods that operate on View and ConstView of the TrackSoA, and cannot be class methods.
+
+template <typename TrackerTraits>
+struct TracksUtilities {
+  using TrackSoAView = typename TrackSoA<TrackerTraits>::template TrackSoALayout<>::View;
+  using TrackSoAConstView = typename TrackSoA<TrackerTraits>::template TrackSoALayout<>::ConstView;
+  using hindex_type = typename TrackSoA<TrackerTraits>::hindex_type;
+
+  // State at the Beam spot
+  // phi,tip,1/pt,cotan(theta),zip
```

Comment:
curious... there is no `pT` method below ?

Reply:
One could directly access the `View` in principle. Could be useful for the sake of uniformity (I'm seeing this in the DQM analyzers where you have to call or not to call the `helper` depending on the variable).

Reply:
By thw way, reading the code I had the impression that what is stored is actually (B field over pT), not (1/pT) ?